### ISSUE TITLE
`mach_port_t` and `thread_t` have different types

### ISSUE THREAD

**Issue by dcuddeback** [fitzgen/mach#2]
I'm working on a PR to move some mach-related types from IOKit-sys (https://github.com/dcuddeback/iokit-sys/blob/f82395b1603b3703da83bfa311f800e88ba96593/src/mach.rs) into this crate. For the most part, this is going pretty smoothly, but I ran into two conflicts:
1. I have `mach_port_t` defined as `c_uint`, but the mach crate defines it as a pointer to an `ipc_port` struct. I determined `c_uint` from reading the header files:
   - from `sys/_types/_mach_port_t.h`:
   
   ``` c
   typedef __darwin_mach_port_t mach_port_t;
   ```
   - from `sys/_types.h`:
   
   ``` c
   typedef __darwin_natural_t __darwin_mach_port_name_t;
   typedef __darwin_mach_port_name_t __darwin_mach_port_t;
   ```
   - from `i386/_types.h`:
   
   ``` c
   typedef unsigned int      __darwin_natural_t;
   ```
2. I have `thread_t` defined as `mach_port_t`, but the mach crate defines it as a pointer to a `thread` struct. I got `mach_port_t` from `mach/mach_types.h`:
   
   ``` c
   typedef mach_port_t       thread_t;
   ```

With a couple of code changes, my IOKit-sys example compiles and runs fine, but I'm curious how we arrived at different types. Also, if I change `mach_port_t` and `thread_t` to `c_uint` and `mach_port_t` respectively, the `dump_process_registers` binary compiles and runs correctly.

Where does `ipc_port` and the `thread` struct come from? I don't see them in the header files, but I'd like to understand how the types in the mach crate were determined so that I can make a compatible PR to add the types needed for IOKit.

**Comment by fitzgen** [comment]
Hi! Thanks for reviving this!

It could be that we are looking at different versions of the headers? I'm not sure which ones I was looking at anymore :-/

**Comment by dcuddeback** [comment]
@fitzgen I might have figured it out. I found this header file online, which defines `mach_port_t` as a pointer to an `ipc_port` struct: http://www.opensource.apple.com/source/xnu/xnu-792.13.8/osfmk/mach/port.h. This looks like how it's defined in this crate, but it doesn't match the header file that I see in XCode. I noticed that the header file I linked to guards the definition with `#ifdef KERNEL`. In the `#else` section, it defined `mach_port_t` as `mach_port_name_t`, which resolves to `unsigned int`. My guess (and it's totally a guess) is that `ipc_port` pointer is for use inside the kernel, and `unsigned int` is for use from userland code.
